### ISSUE TITLE
404 Response from promise on `app.charge(...)` and `app.payout(...)`

### ISSUE THREAD

**Issue by codehakase** [mobius-network/mobius-client-js#23]
Hello Team!

The SDK throws a `Request failed with status code 400` (from all possible ways of catching the error, that message is still returned` This error is thrown sometimes on the `app.charge(...)` and `app.payout(...)` methods.

Here's a snippet of how I'm calling each methods:

```javascript
const handler = async () => {
 try { // parent
  ...
  let dapp;
  try {
   dapp = await MobiusClient.AppBuilder.build(keypair, userPublicKey);
  } catch (err) {
   if (err) {
    throw new Error(err.message);
   }
  }
  // process payout
  try {
   await dapp.payout(amount); // or await dapp.charge(amount)...
   console.log(`New payout of ${amount} made to ${userPublicKey}`);
  } catch (err) {
   if (err) throw new Error(`could not make payout to user: ${err.message}`);
  }
 }
 catch (err) {
  if (err) {
   // custom error handling here
  }
 }
}

```

**Comment by mrtpain** [comment]
In `dapp = await MobiusClient.AppBuilder.build(keypair, userPublicKey);` , `keypair` is supposed to be the dev secret seed not the full keypair. I can't see that variable being assigned to that but could be your problem.

**Comment by nebolsin** [comment]
@codehakase can you please share the `appPublicKey`, `userPublicKey` and `amount` which end up with this failure? Your code looks pretty standard, so I'd guess this error has something to do with actual accounts setup on the network.

**Comment by codehakase** [comment]
@mrtpain `keypair` is actually the dev secret key, I abstracted the assignment.

**Comment by codehakase** [comment]
@nebolsin the error exists on different wallets (dapp store instances), here's one we used more for testing on public net:

`userPublicKey`: GAC5JPULSLGW3Y2IVFFAAVH2V2KRDWZCVGRXLRS53HQUJYPTYDKA5ADM
`appPublicKey`:  GC7KGS7MJ3EFYLAG4UQFPAW3CDBHQNEP55FJJWMBZVU2F3LR6AI5W4E5
`amount`: different payouts were expected to complete but failed: `200, 20, 5, 10, 1 MOBI`

**Comment by codehakase** [comment]
The `400 error` is thrown at this point:

```javascript
...
// process payout
  try {
   await dapp.payout(amount); // or await dapp.charge(amount)...
   console.log(`New payout of ${amount} made to ${userPublicKey}`);
  } catch (err) {
   // error is thrown here: `could not make payout to user: Request failed with status code 400`
   if (err) throw new Error(`could not make payout to user: ${err.message}`); 
  }
```

**Comment by nebolsin** [comment]
@codehakase If you experience this on pubnet, another thing to check is that you correctly configured the SDK to use the network, using the following code:

```
const mobius = new Mobius.Client();
mobius.network = StellarSdk.Networks.PUBLIC;
```

Using `StellarSdk.Network.usePublicNetwork()` *is not enough* and will actually result in 400 errors because the app will still use testnet's Horizon server but generate signatures for public network.

**Comment by codehakase** [comment]
@nebolsin Oh... I see, that may just be the case, will update my code and run some tests

**Comment by codehakase** [comment]
@nebolsin based on your last reply, just for heads up, here's how I initialise the mobius client in use by each handler:

```javascript
...
const MobiusClient = require('@mobius-network/mobius-client-js');

// dapp instance is created with
dapp = await MobiusClient.AppBuilder.build(keypair, userPublicKey);
```

I can't really picture where to explicitly set the network on the mobius instance in this scenario.

**Comment by mrtpain** [comment]
@codehakase index.js or where you set CORS, you can see the Flappy Dapp example [here](https://github.com/mobius-network/flappy-dapp/blob/master/backend/nodejs-api/index.js)

**Comment by codehakase** [comment]
@mrtpain thanks!

**Comment by codehakase** [comment]
@nebolsin I updated my code to set the network on the client, and on various runs, the same error is thrown from both the `charge` and `payout` functions

**Comment by mrtpain** [comment]
@codehakase I would suggest setting a breakpoint using a debugger and seeing the full details of the error. The 400 response is likely coming from horizon.

**Comment by CashMoneyLumens** [comment]
Guys it’s mikeways here this is my project anyone who could help it be much appreactied I have limted coding knowlodge ... if someone could help me understand the issue that I’m having... or point me in the right direction

**Comment by CashMoneyLumens** [comment]
So mr tpain your saying horizon is throwing the error code ... not it’s a problem with horizon itself ?

**Comment by codehakase** [comment]
@CashMoneyLumens we are working on a fix, this is just to let the mobius team note if the issue if with the sdk itself

**Comment by mrtpain** [comment]
I don't believe the issue is with the SDK itself, especially since it hasn't had an update in a few months and is working fine elsewhere but I'm not ruling it out. 

The error being thrown is `Transaction Failed` and `tx_bad_seq` which is a bad sequence number. This error can be thrown for a few different reasons. 

@codehakase I noticed the stack trace was done with a transaction on testnet, so I assume that the issue is not only on public network as mentioned before. 

I will dig into this more and get back with findings now that we know more about the error.

**Comment by codehakase** [comment]
Yeah, I got same for both public and testnet.

Thanks

**Comment by codehakase** [comment]
@nebolsin this is a valid point. Majority of the transaction on [ GC7K...5W4E5](https://steexp.com/account/GC7KGS7MJ3EFYLAG4UQFPAW3CDBHQNEP55FJJWMBZVU2F3LR6AI5W4E5#transactions) are `charge` operations, which were sent in as multiple requests hitting the specific handler.

**Comment by nebolsin** [comment]
@codehakase I think with the current version of SDK if you expect multiple concurrent operations on the same user account, you need to implement either external synchronization or retry on the app level, depending on your use case. If retry will work for your scenario, it’s probably the simplest solution to start with (detect this error in the catch block and attempt to execute the failed operation again)

**Comment by codehakase** [comment]
@nebolsin thanks, exactly what I had in mind

**Comment by codehakase** [comment]
So one can close this issue?
